Updates to onchip_memory_cache tutorial

mdbtucker opened the pull request:
## Description

Preliminary review for rewrite of the onchip_memory_cache tutorial.

## Type of change

Please delete options that are not relevant. Add a 'X' to the one that is applicable. 

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Implement fixes for ONSAM Jiras

## How Has This Been Tested?
 - running regtests to confirm performance at various cache depths 
 - cache depth 0 (II2 with load latency 1): https://spetc.intel.com/testsummary?testRunIds=6561034
 - cache depth 1 (II1 with load latency 1): https://spetc.intel.com/testsummary?testRunIds=6561057
 - cache depth 7 (II1 with load latency 4): https://spetc.intel.com/testsummary?testRunIds=6561065
 - testing result: FMAX shows no change or slight improvement of this technique over ivdep (S10 with cache ~5% better FMAX, S10 without cache and A10 with and without cache basically unchanged)
 - ran Linux regtest with default parameters, all tests passed
 - running Windows regtests on emulation, a10_ref and s10_ref, waiting for results (had some issues launching the test)

mdbtucker committed:
Added array specifier to ivdep attribute

mdbtucker committed:
Updated onchip_memory_cache tutorial to use new CachedLocalMemory class.  Various tutorial upgrades, including testing across multiple cache sizes.  Current version does not acheive II1 due to compiler bug when array name is specified.

mdbtucker committed:
Removed array specifier on ivdep as it results in the ivdep not working as desired

mdbtucker committed:
Fixed typo in status message

Signed-off-by: mtucker <mike.d.b.tucker@intel.com>

mdbtucker committed:
Replaced class that implements local memory with cache to class that only implements the cache

Signed-off-by: mtucker <mike.d.b.tucker@intel.com>

mdbtucker committed:
Safe ivdep cache implemented, reports look good and works in emulation.

mdbtucker committed:
Fully encapsulated the local memory with cache into a class.  User code now very simple, and no ivdep required.  Cache depth for good performance is 2x what is needed with ivdep, and valid bit needed per entry, which means more logic as well as more registers.

mdbtucker committed:
Pull from end of cache instead of middle of cache to write to memory.  Results in much smaller cache size required.

mdbtucker committed:
Added ability to specify max and min cache depth to explore, to allow compiling with a single cache depth for fmax exploration.  Fixed a comment in unrolled_loop.hpp.

mdbtucker committed:
Updated stdout messages

mdbtucker committed:
More updates to stdout messages

mdbtucker committed:
removed fpga_register attribute from class member variables, as it was being ignored by the compiler and causing a warning

yuguen commented on .gitignore:
```diff
@@ -1,5 +1,5 @@
 # CMAKE build directories
-build/
+build*/
```
This is definitely helpful, but:
1/ This is outside the FPGA part of the repo
2/ Not really related to the onchip_memory_cache tutorial
Should this be in a different PR?

mdbtucker replied:
Agreed.  Reverted this change.

yuguen commented on DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/onchip_memory_cache/src/CMakeLists.txt:
```diff
@@ -19,16 +19,32 @@ endif()
+set(EMULATOR_LINK_FLAGS "${AC_TYPES_FLAG} -fintelfpga")
+set(HARDWARE_COMPILE_FLAGS "-Wall ${WIN_FLAG} ${CACHE_DEPTH_FLAG} ${AC_TYPES_FLAG} -fintelfpga")
+set(REPORT_LINK_FLAGS "-fintelfpga -Xshardware -Xsboard=${FPGA_BOARD} ${CACHE_DEPTH_FLAG} ${USER_HARDWARE_FLAGS}")
+set(HARDWARE_LINK_FLAGS "${REPORT_LINK_FLAGS} ${AC_TYPES_FLAG}")
```
I think you should use the AC_TYPES_FLAG for the report as well. This way you would only have REPORT_LINK_FLAGS and not HARDWARE_LINK_FLAGS, which would be more similar to the other designs cmake files

mdbtucker replied:
When -qactypes is added to the REPORT_LINK_FLAGS, it generates a warning:

/p/psg/swip/w/mtucker/oneapi_samples_mdbtucker.1/DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/onchip_memory_cache/build$ make report
Scanning dependencies of target onchip_memory_cache_report.a
[ 50%] Building CXX object src/CMakeFiles/onchip_memory_cache_report.a.dir/onchip_memory_cache.cpp.o
[100%] Linking CXX executable ../onchip_memory_cache_report.a
dpcpp: warning: argument unused during compilation: '-qactypes' [-Wunused-command-line-argument]

That's why I only add it when compiling the fpga target.

yuguen commented on DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/onchip_memory_cache/src/onchip_memory_cache.cpp:
```diff
@@ -5,140 +5,97 @@
 // e.g., $ONEAPI_ROOT/dev-utilities//include/dpc_common.hpp
 #include "dpc_common.hpp"
 
-using namespace sycl;
```
Thanks for making this change as part of this work!

mdbtucker replied:
You're welcome :-)

yuguen commented on DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/onchip_memory_cache/src/onchip_memory_cache.cpp:
```diff
@@ -5,140 +5,97 @@
-        } else {
-          output[b] = local_output_with_cache[b];
-        }
+        output[b] = histogram.read(b);
```
You did a similar change below: you can rename `b` to `hist_group`

mdbtucker replied:
Replaced all uses of b with hist_group

yuguen commented on DirectProgramming/DPC++FPGA/include/cached_local_memory.hpp:
```diff
@@ -0,0 +1,124 @@
+      cache_valid_[i] = false;
+    }
+  }
+  CachedLocalMemory(T init_val) {
```
What is the point of having an initializer if all the valids are false?

mdbtucker replied:
Setting the valids to false just means you ignore the cache.  The initializer sets all the values in the actual memory.  So for the histogram example, I initialize the memory to 0, which is required for correct operation.

yuguen replied:
Why it is required? 
If the valids are 0, then no matter what, the data at that location is never going to be read
So it does not matter if it is initialized or not, does it?

mdbtucker replied:
There is only a valid per cache entry.  If all the cache entries are invalid, then a read will pull data from the actual memory.  If the data in the actual memory has not been initialized, then you read garbage.

yuguen replied:
Oh right, sorry for some reason I was thinking that the initialization was for the cache values! My bad!

yuguen commented on DirectProgramming/DPC++FPGA/include/unrolled_loop.hpp:
```diff
@@ -87,7 +87,7 @@ constexpr void UnrolledLoop(F&& f) {
 
 //
-// Convenience implementation (C)
+// Convenience implementation (D)
```
What this intended?

mdbtucker replied:
Yes it was.  It's not really related to this change, just something I noticed while working on this change, a simple error in the comments of this file.  I could put it into a separate PR if you want, but it's small enough I was hoping to just sneak it through with this one . . .

yuguen replied:
I'm fine with the sneaking :)

yuguen commented on DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/onchip_memory_cache/src/onchip_memory_cache.cpp:
```diff
@@ -157,70 +114,70 @@ int main() {
+      if (i == 0) { std::cout << " (no cache)"; }
+      std::cout << std::endl;
+
+      fpga_tools::UnrolledLoop<kMinCacheDepth, kMaxCacheDepth+1>([&](auto j) {
```
Interesting workaround.
This is not an obvious technique that we expect the user to understand (and the underlying reasons) just by having a look at the code of this tutorial. 
I think you should add a comment explaining why you did that, and that it has nothing to do with how you should use the histogram

mdbtucker replied:
Good point, I added a comment.  Have a look and please resolve this comment if you think it's sufficient.

yuguen commented:
You probably though about it already, but there are _two_ readmes to update: the tutorial readme and the include readme :)

yuguen commented:
You also probably need to update the MSVS files to have the added include path to the include folder

mdbtucker committed:
Reverted change to .gitignore file

Signed-off-by: mtucker <mike.d.b.tucker@intel.com>

mdbtucker committed:
Changed all uses of variable name b to hist_group

Signed-off-by: mtucker <mike.d.b.tucker@intel.com>

mdbtucker committed:
Added comment to explain the use of UnrolledLoop

Signed-off-by: mtucker <mike.d.b.tucker@intel.com>

mdbtucker commented:
> You probably though about it already, but there are _two_ readmes to update: the tutorial readme and the include readme :)

Of course I thought of that!  Why I'm offended at the very implication that I might have forgotten!  How dare you sir?  HOW DARE YOU?!?

(ok, maybe I forgot a tiny bit)

mdbtucker committed:
Rename CachedLocalMemory to OnchipMemoryWithCache

mdbtucker commented:
I spoke with Joe about using the term 'local memory', and he pointed out it's overly restrictive (this technique can apply to private memory, and to device globals).  We agreed that the term 'onchip memory' is more accurate.

I still feel that 'onchip_memory_cache' is a bit ambiguous - it could mean just the cache is in onchip memory and the memory itself is off chip, or it could mean the memory BEING cached is on chip.  So I've renamed my class to 'OnchipMemoryWithCache', and renamed the header file to match.

I feel like this is close enough to 'onchip_memory_cache' that I don't have to rename the tutorial.  I think renaming the tutorial is a bit of a pain and has some trickle-down extra work involved that I'd rather avoid.  Let me know if you don't like this compromise.

yuguen commented:
> I spoke with Joe about using the term 'local memory', and he pointed out it's overly restrictive (this technique can apply to private memory, and to device globals). We agreed that the term 'onchip memory' is more accurate.
> 
> I still feel that 'onchip_memory_cache' is a bit ambiguous - it could mean just the cache is in onchip memory and the memory itself is off chip, or it could mean the memory BEING cached is on chip. So I've renamed my class to 'OnchipMemoryWithCache', and renamed the header file to match.
> 
> I feel like this is close enough to 'onchip_memory_cache' that I don't have to rename the tutorial. I think renaming the tutorial is a bit of a pain and has some trickle-down extra work involved that I'd rather avoid. Let me know if you don't like this compromise.

I think that this is a very good compromise

mdbtucker committed:
Updated README files.

Signed-off-by: mtucker <mike.d.b.tucker@intel.com>

mdbtucker commented:
@yuguen-intel I've updated the README files (just need to do a HW run to capture the expected output), could you take a look when you have time?

mdbtucker committed:
Updated json file to support the include directory

Signed-off-by: mtucker <mike.d.b.tucker@intel.com>

mdbtucker committed:
Added include directory to vcxproj file

Signed-off-by: mtucker <mike.d.b.tucker@intel.com>

mdbtucker commented:
> You also probably need to update the MSVS files to have the added include path to the include folder

I've updated the vcxproj file to add the include directory, but since regtests don't currently support testing the MSVC files, I have no way to test this.

I also updated the .json file to suppor the include directory, but since the CI is currently not working, I also have no way to test this change.

whitepau commented on DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/onchip_memory_cache/README.md:
```diff
@@ -21,36 +21,36 @@ In DPC++ task kernels for FPGA, our objective is to achieve an initiation interv
 When the loop contains a loop-carried variable implemented in on-chip memory, the compiler often *cannot* achieve II=1 because the memory access takes more than one clock cycle. If the updated memory location may be needed on the next loop iteration, the next iteration must be delayed to allow time for the update, hence II > 1.
 
-The on-chip memory cache technique breaks this dependency by storing recently-accessed values in a cache capable of a 1-cycle read-modify-write operation. The cache is implemented in FPGA registers rather than on-chip memory. By pulling memory accesses preferentially from the register cache, the loop-carried dependency is broken.
+The technique of using the OnchipMemoryWithCache class breaks this dependency by storing recently-accessed values in a cache capable of a 1-cycle read-modify-write operation. The cache is implemented in FPGA registers rather than on-chip memory. By pulling memory accesses preferentially from the register cache, the loop-carried dependency is broken.
```
```suggestion
The technique of using the `OnchipMemoryWithCache` class breaks this dependency by storing recently-accessed values in a cache capable of a 1-cycle read-modify-write operation. The cache is implemented in FPGA registers rather than on-chip memory. By pulling memory accesses preferentially from the register cache, the loop-carried dependency is broken.
```

mdbtucker replied:
Fixed!

whitepau commented on DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/onchip_memory_cache/README.md:
```diff
@@ -21,36 +21,36 @@ In DPC++ task kernels for FPGA, our objective is to achieve an initiation interv
 
 ***Failure to achieve II=1 because of a loop-carried memory dependency in on-chip memory***:
-The on-chip memory cache technique is applicable if the compiler could not pipeline a loop with II=1 because of an on-chip memory dependency. (If the compiler could not achieve II=1 because of a *global* memory dependency, this technique does not apply as the access latencies are too great.)
+The on-chip memory with cache technique is applicable if the compiler could not pipeline a loop with II=1 because of an on-chip memory dependency. (If the compiler could not achieve II=1 because of a *global* memory dependency, this technique does not apply as the access latencies are too great.)
```
I think this would benefit from a small graphic showing how the load/store latency of the memory gets on the loop-carried feedback path.

mdbtucker replied:
I'm not sure what such a diagram would look like, and I'm afraid I've spent more time on this already than I had planned, so that will have to wait for a future update (perhaps by you?)

whitepau commented on DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/onchip_memory_cache/README.md:
```diff
@@ -21,36 +21,36 @@ In DPC++ task kernels for FPGA, our objective is to achieve an initiation interv
 
 ***Failure to achieve II=1 because of a loop-carried memory dependency in on-chip memory***:
-The on-chip memory cache technique is applicable if the compiler could not pipeline a loop with II=1 because of an on-chip memory dependency. (If the compiler could not achieve II=1 because of a *global* memory dependency, this technique does not apply as the access latencies are too great.)
+The on-chip memory with cache technique is applicable if the compiler could not pipeline a loop with II=1 because of an on-chip memory dependency. (If the compiler could not achieve II=1 because of a *global* memory dependency, this technique does not apply as the access latencies are too great.)
```
I assume that when IP Authoring is no longer a mischievous glimmer in the eyes of our engineers, we can generalize this from "on-chip memory" to "any *fixed-latency* memory accesses"? Functionally speaking, I realize that these are the same, but in practice this means that kernels that read-modify-write to a fixed-latency MM Host interface could also benefit from this design pattern.

mdbtucker replied:
I guess the technique could work, but this abstracted class would not, as it contains the memory itself.

whitepau commented on DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/onchip_memory_cache/README.md:
```diff
@@ -21,36 +21,36 @@ In DPC++ task kernels for FPGA, our objective is to achieve an initiation interv
-The on-chip memory cache technique is applicable if the compiler could not pipeline a loop with II=1 because of an on-chip memory dependency. (If the compiler could not achieve II=1 because of a *global* memory dependency, this technique does not apply as the access latencies are too great.)
+The on-chip memory with cache technique is applicable if the compiler could not pipeline a loop with II=1 because of an on-chip memory dependency. (If the compiler could not achieve II=1 because of a *global* memory dependency, this technique does not apply as the access latencies are too great.)
 
 To check this for a given design, view the "Loops Analysis" section of its optimization report. The report lists the II of all loops and explains why a lower II is not achievable. Check whether the reason given resembles "the compiler failed to schedule this loop with smaller II due to memory dependency". The report will describe the "most critical loop feedback path during scheduling". Check whe ...
```
Is there a way we can put a screenshot in here? I REALLY don't like describing what the report looks like in words when a picture would remove ambiguity. I know pictures are tough to maintain because when the GUI changes the picture needs to be re-screenshot, but that should be easy to do since these code samples are all regtested.

mdbtucker replied:
I don't think we have screenshots of reports in any other code sample.  I agree with you it's preferable to the user, but I think the burden of keeping the screen shots up to date is excessive.

mdbtucker committed:
Only print the Verification Passed message once, not once per kernel

Signed-off-by: mtucker <mike.d.b.tucker@intel.com>

mdbtucker committed:
Replace ulong with unsigned long, as ulong is unrecognized when compiling for Windows

Signed-off-by: mtucker <mike.d.b.tucker@intel.com>

yuguen commented:
> > You also probably need to update the MSVS files to have the added include path to the include folder
> 
> I've updated the vcxproj file to add the include directory, but since regtests don't currently support testing the MSVC files, I have no way to test this.

Yeah it's quite a pain, but if you look at the qrd one for example, you should have the correct way of doing it. I did test the MSVS files for the qrd on my laptop install of oneapi back then

> 
> I also updated the .json file to suppor the include directory, but since the CI is currently not working, I also have no way to test this change.

~~This will be tested when you'll run the regtest~~ You are right, as long as the CI is down we won't be able to test this

yuguen commented on DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/onchip_memory_cache/README.md:
```diff
@@ -21,36 +21,36 @@ In DPC++ task kernels for FPGA, our objective is to achieve an initiation interv
 
 ***Failure to achieve II=1 because of a loop-carried memory dependency in on-chip memory***:
-The on-chip memory cache technique is applicable if the compiler could not pipeline a loop with II=1 because of an on-chip memory dependency. (If the compiler could not achieve II=1 because of a *global* memory dependency, this technique does not apply as the access latencies are too great.)
+The on-chip memory with cache technique is applicable if the compiler could not pipeline a loop with II=1 because of an on-chip memory dependency. (If the compiler could not achieve II=1 because of a *global* memory dependency, this technique does not apply as the access latencies are too great.)
```
Just a general question: can't we use a very large cache in that case? (e.g. 300 entries)

mdbtucker replied:
We could, but I'm pretty sure it would not be efficient/performant to create a 300 entry cache with this method (purely register backed).  A cache for off-chip memory would probably need a combination of local memory and registers in the cache.

yuguen commented on DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/onchip_memory_cache/sample.json:
```diff
@@ -5,9 +5,18 @@
   "os": ["linux", "windows"],
-  "builder": ["ide", "cmake"],
   "targetDevice": ["FPGA"],
+  "builder": ["ide", "cmake"],
```
I'm not sure if it makes a difference, but you moved the builder entry one line down.
It should not make a difference, but to avoid braking a flow somewhere, we might as well let it where it was before

mdbtucker replied:
I did this to match the sample.json I was using as a reference (pipe_array), which I know passes the CI.

yuguen replied:
Understood, then that's fine

yuguen commented on DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/onchip_memory_cache/sample.json:
```diff
@@ -5,9 +5,18 @@
+    "include": [
+      "Tutorials/DesignPatterns/onchip_memory_cache",
+      "include/onchip_memory_with_cache.hpp",
+      "include/unrolled_loop.hpp"
```
Looking at what is done in qrd, it only includes "include" rather than individual files.
As we can't test this file right now, I'd suggest doing the same as what we know worked before here

yuguen replied:
Looking at the merge sort design, it lists the required include files, so you are all good

mdbtucker committed:
Added output from S10 hardware run to the README

Signed-off-by: mtucker <mike.d.b.tucker@intel.com>

mdbtucker committed:
Removed TODO from the README

Signed-off-by: mtucker <mike.d.b.tucker@intel.com>

yuguen reviewed (approved):
Looks good to me :)

whitepau commented on DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/onchip_memory_cache/README.md:
```diff
@@ -21,36 +21,36 @@ In DPC++ task kernels for FPGA, our objective is to achieve an initiation interv
+For user designs, each iteration takes only a few moments to compile the reports. It is important to find the *minimal* value of `CACHE_DEPTH` that results in a maximal performance increase. Unnecessarily large values of `CACHE_DEPTH` consume unnecessary FPGA resources and can reduce f<sub>MAX</sub>. Therefore, at a `CACHE_DEPTH` that results in II=1 and load latency = 1, if further increases to  ...
 
-Two versions of the histogram kernel are implemented in the tutorial: one with and one without caching. The report shows II > 1 for the loop in the kernel without caching and II = 1 for the one with caching.
+This tutorial creates multiple kernels sweeping across different cache depths within a single design.  This allows a single compile of the reports to determine the optimal cache depth.
```
this would be a great place for a graph :)

mdbtucker replied:
It would be a pretty boring graph, since the throughput is the same for all the kernels except the one without a cache.

whitepau commented on DirectProgramming/DPC++FPGA/include/onchip_memory_with_cache.hpp:
```diff
@@ -0,0 +1,124 @@
+    bool in_cache = false;
+
+    // check the cache, newest value will take precedence
+    #pragma unroll
```
do we not have an `[[intel::unroll]]` attribute?

mdbtucker replied:
Hmm, I never thought of that, but it's a perfectly reasonable question.  As far as I can tell, the answer is no such attribute exists.  If it does exist, it's very hard to find in our docs, and the loop_unroll code sample does not mention it.

I've reached out to the loops team to ask WHY that is the case, I will let you know the answer (if there is one).

whitepau commented on DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/onchip_memory_cache/src/onchip_memory_cache.cpp:
```diff
@@ -5,140 +5,97 @@
-
+      // A k_cache_depth of 0 is equivalent to a standard array with no cache
+      fpga_tools::OnchipMemoryWithCache<uint32_t, kNumOutputs, k_cache_depth> 
+        histogram(0);
```
can you call this `histogram_mem` so it's easier to see later in the code that it's a memory? (I'm stupid and need to be beat over the head with information so it sinks in.)
```suggestion
        histogram_mem(0);
```

mdbtucker replied:
It's not just a memory though, it's a memory with a cache.

Also, I don't think it's standard practice to add _mem to anything that's a memory.  Most arrays aren't called, for example, q_matrix_mem or steering_vector_mem.  I don't think this would enhance the clarity of the code.

whitepau commented on DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/onchip_memory_cache/src/onchip_memory_cache.cpp:
```diff
@@ -5,140 +5,97 @@
-            last_sum_index[i] = last_sum_index[i + 1];
-          }
-        }
+      for (uint32_t n = 0; n < kInitNumInputs; ++n) {
```
Why not use `[[intel::iteration_interval 1]]` to force this loop to achieve II=1? Or is the problem that it doesn't achieve II=1 because you can't use `ivdep` safely as has been discussed ad nauseum?

mdbtucker replied:
It can't achieve II1 for the case with 0 depth cache (no cache).  So I can't add an II1 here.

whitepau commented on DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/onchip_memory_cache/src/onchip_memory_cache.cpp:
```diff
@@ -5,140 +5,97 @@
+    sycl::accessor output(output_buf, h, sycl::write_only, sycl::no_init);
 
-    h.single_task<Task<use_cache>>([=]() [[intel::kernel_args_restrict]] {
+    h.single_task<HistogramID<k_cache_depth>>(
```
this is a nice clean kernel.

mdbtucker replied:
I've got global memory read and write pointers that don't alias, so I think the kernel_args_restrict is appropriate here.

whitepau commented on DirectProgramming/DPC++FPGA/include/onchip_memory_with_cache.hpp:
```diff
@@ -0,0 +1,124 @@
+  // explicitly communicate to developers that we don't want to support a
+  // square bracket operator that returns a reference, as it would allow
+  // modification of the memory without updating the cache
+  template <typename I> T& operator[](I addr) = delete;
```
Why not implement a square-bracket operator that simply calls the `write()` function like you did for the `read()`? I think `[]` is a better API than `read()` and `write()`.

mdbtucker replied:
Normally a square bracket operator returns a reference.  I can't do that here, I need a function call to read, and a different function call to write.  I don't know if that's possible with a singler operator, basically I need to do different things depending on if it's being used on the LHS or RHS of an operation.  If it is possible, it's beyond my C++ skillz.

mdbtucker committed:
Minor readme update

Signed-off-by: mtucker <mike.d.b.tucker@intel.com>
